### ISSUE TITLE
`ReduceMax` before `Softmax` ignored during conversion

### ISSUE THREAD

**Issue by H4dr1en** [PINTO0309/onnx2tf#182]
### Issue Type

Feature Request, Others

### onnx2tf version number

1.6.3

### onnx version number

1.13.0

### tensorflow version number

2.10.0

### Download URL for ONNX

https://drive.google.com/file/d/1HthOr18Udxi7dUQnMCdPeWiTvvy1iMOV/view?usp=share_link

### Parameter Replacement JSON

```json
{}
```

### Description

1. Product development
2. I have a model producing overflows when running a softmax layer. In pytorch the softmax operation doesn't produce overflows (I guess it's using a numerically stable version like `softmax(x-max(x))`), but when running the converted model on device I get the error: 

```
A/libc: Fatal signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x20 in tid 28138 (pool-17-thread-)
```

So I tried to replace the softmax layer with `softmax(x-max(x))`: 

<img width="251" alt="Screenshot 2023-02-10 at 15 00 50" src="https://user-images.githubusercontent.com/16240134/218111296-92d4ea8e-9ab7-4468-acb5-48a26e03c723.png">

But onnx2tf seems to ignore it during the conversion:

<img width="206" alt="Screenshot 2023-02-10 at 15 01 05" src="https://user-images.githubusercontent.com/16240134/218111404-aebbb506-8933-4801-ae81-208084822358.png">

3. I tried to run onnx2tf with `not_use_onnxsim=True` without succes
4. This bug prevents me from running the model on device

**Comment by PINTO0309** [comment]
Thank you. It is a rather strange situation. The conversion logs all appear to be correct, and I actually debugged the program line by line and it appears to be building the model correctly. However, the final output model is broken.

So far I have not been able to examine the problem areas.

```
INFO: input_op_name: image shape: [1, 23, 64, 48] dtype: float32

INFO: onnx_op_type: Reshape onnx_op_name: Reshape_1
INFO:  input_name.1: image shape: [1, 23, 64, 48] dtype: float32
INFO:  input_name.2: onnx::Reshape_426 shape: [3] dtype: <class 'numpy.int64'>
INFO:  output_name.1: onnx::ReduceMax_427 shape: [1, 23, 3072] dtype: float32
INFO: tf_op_type: reshape
INFO:  input.1.tensor: name: tf.compat.v1.transpose/transpose:0 shape: (1, 23, 64, 48) dtype: <dtype: 'float32'> 
INFO:  input.2.shape: val: [1, -1, 3072] 
INFO:  output.1.output: name: tf.reshape/Reshape:0 shape: (1, 23, 3072) dtype: <dtype: 'float32'> 

INFO: onnx_op_type: ReduceMax onnx_op_name: ReduceMax_2
INFO:  input_name.1: onnx::ReduceMax_427 shape: [1, 23, 3072] dtype: float32
INFO:  output_name.1: onnx::Unsqueeze_428 shape: [1, 23] dtype: float32
INFO: tf_op_type: reduce_max
INFO:  input.1.axis0: val: 2 
INFO:  input.2.input_tensor: name: tf.reshape/Reshape:0 shape: (1, 23, 3072) dtype: <dtype: 'float32'> 
INFO:  input.3.keepdims: val: False 
INFO:  output.1.output: name: tf.math.reduce_max/Max:0 shape: (1, 23) dtype: <dtype: 'float32'> 

INFO: onnx_op_type: Unsqueeze onnx_op_name: Unsqueeze_4
INFO:  input_name.1: onnx::Unsqueeze_428 shape: [1, 23] dtype: float32
INFO:  input_name.2: onnx::Unsqueeze_429 shape: [1] dtype: <class 'numpy.int64'>
INFO:  output_name.1: onnx::Sub_430 shape: [1, 23, 1] dtype: float32
INFO: tf_op_type: reshape
INFO:  input.1.tensor: name: tf.math.reduce_max/Max:0 shape: (1, 23) dtype: <dtype: 'float32'> 
INFO:  input.2.shape: val: [1, 23, 1] 
INFO:  output.1.output: name: tf.reshape_1/Reshape:0 shape: (1, 23, 1) dtype: <dtype: 'float32'> 

INFO: onnx_op_type: Sub onnx_op_name: Sub_5
INFO:  input_name.1: onnx::ReduceMax_427 shape: [1, 23, 3072] dtype: float32
INFO:  input_name.2: onnx::Sub_430 shape: [1, 23, 1] dtype: float32
INFO:  output_name.1: flat shape: [1, 23, 3072] dtype: float32
INFO: tf_op_type: subtract
INFO:  input.1.x: name: tf.reshape/Reshape:0 shape: (1, 23, 3072) dtype: <dtype: 'float32'> 
INFO:  input.2.y: name: tf.expand_dims_1/ExpandDims:0 shape: (1, 23, 1) dtype: <dtype: 'float32'> 
INFO:  output.1.output: name: tf.math.subtract/Sub:0 shape: (1, 23, 3072) dtype: <dtype: 'float32'> 

INFO: onnx_op_type: Softmax onnx_op_name: Softmax_6
INFO:  input_name.1: flat shape: [1, 23, 3072] dtype: float32
INFO:  output_name.1: test shape: [1, 23, 3072] dtype: float32
INFO: tf_op_type: softmax_v2
INFO:  input.1.logits: name: tf.math.subtract/Sub:0 shape: (1, 23, 3072) dtype: <dtype: 'float32'> 
INFO:  input.2.axis: val: 2 
INFO:  output.1.output: name: tf.nn.softmax_1/Softmax:0 shape: (1, 23, 3072) dtype: <dtype: 'float32'> 

Model: "model_1"
____________________________________________________________________________________________________________________________________________
 Layer (type)                                 Output Shape                   Param #         Connected to                                   
============================================================================================================================================
 image (InputLayer)                           [(1, 64, 48, 23)]              0               []                                             
                                                                                                                                            
 tf.compat.v1.transpose (TFOpLambda)          (1, 23, 64, 48)                0               ['image[0][0]']                                
                                                                                                                                            
 tf.reshape (TFOpLambda)                      (1, 23, 3072)                  0               ['tf.compat.v1.transpose[0][0]']               
                                                                                                                                            
 tf.math.reduce_max (TFOpLambda)              (1, 23)                        0               ['tf.reshape[0][0]']                           
                                                                                                                                            
 tf.reshape_1 (TFOpLambda)                    (1, 23, 1)                     0               ['tf.math.reduce_max[0][0]']                   
                                                                                                                                            
 tf.compat.v1.squeeze (TFOpLambda)            (23,)                          0               ['tf.reshape_1[0][0]']                         
                                                                                                                                            
 tf.expand_dims (TFOpLambda)                  (23, 1)                        0               ['tf.compat.v1.squeeze[0][0]']                 
                                                                                                                                            
 tf.expand_dims_1 (TFOpLambda)                (1, 23, 1)                     0               ['tf.expand_dims[0][0]']                       
                                                                                                                                            
 tf.math.subtract (TFOpLambda)                (1, 23, 3072)                  0               ['tf.reshape[0][0]',                           
                                                                                              'tf.expand_dims_1[0][0]']                     
                                                                                                                                            
 tf.nn.softmax_1 (TFOpLambda)                 (1, 23, 3072)                  0               ['tf.math.subtract[0][0]']                     
                                                                                                                                            
============================================================================================================================================
Total params: 0
Trainable params: 0
Non-trainable params: 0
____________________________________________________________________________________________________________________________________________

saved_model output started ==========================================================
saved_model output complete!
WARNING:absl:Please consider providing the trackable_obj argument in the from_concrete_functions. Providing without the trackable_obj argument is deprecated and it will use the deprecated conversion path.
Estimated count of arithmetic ops: 4.522 M  ops, equivalently 2.261 M  MACs
Float32 tflite output complete!
Estimated count of arithmetic ops: 4.522 M  ops, equivalently 2.261 M  MACs
Float16 tflite output complete!

ONNX and TF output value validation started =========================================
INFO: validation_conditions: np.allclose(onnx_outputs, tf_outputs, rtol=0.0, atol=0.0001, equal_nan=True)
INFO: onnx_output_name: onnx::ReduceMax_427 tf_output_name: tf.reshape/Reshape:0 shape: (1, 23, 3072) dtype: float32 validate_result:  Matches 
INFO: onnx_output_name: onnx::Unsqueeze_428 tf_output_name: tf.math.reduce_max/Max:0 shape: (1, 23) dtype: float32 validate_result:  Matches 
INFO: onnx_output_name: onnx::Sub_430 tf_output_name: tf.reshape_1/Reshape:0 shape: (1, 23, 1) dtype: float32 validate_result:  Matches 
INFO: onnx_output_name: flat tf_output_name: tf.math.subtract/Sub:0 shape: (1, 23, 3072) dtype: float32 validate_result:  Matches 
INFO: onnx_output_name: test tf_output_name: tf.nn.softmax_1/Softmax:0 shape: (1, 23, 3072) dtype: float32 validate_result:  Matches 
```

**Comment by PINTO0309** [comment]
Apparently you have found a bug in the TensorFlow optimizer.

- Left: Keras .h5, Right: tflite
![image](https://user-images.githubusercontent.com/33194443/218117556-2efcdc1d-6562-4382-84fd-94b76bec8232.png)

**Comment by PINTO0309** [comment]
Very interesting indeed.
```python
import torch
import torch.nn as nn


class Model(nn.Module):
    def __init__(self):
        super(Model, self).__init__()

    def forward(self, x):
        res = torch.reshape(x, [1,-1,3072])
        workaround_res = res + 1.0
        reduce, _ = torch.max(res, dim=2, keepdim=True)
        workaround_mul = reduce + 1.0
        sub = torch.subtract(workaround_res, workaround_mul)
        sub = sub + 1.0
        ret = torch.softmax(sub, dim=2)
        ret = ret + 1.0
        return ret

model = Model()

x = torch.randn([1,23,64,48])
onnx_file = f'reducemax_softmax_workaround.onnx'
torch.onnx.export(
    model,
    args=(x),
    f=onnx_file,
    opset_version=11,
    input_names=[
        'input',
    ],
    output_names=[
        'output',
    ],
)
import onnx
from onnxsim import simplify
model_onnx2 = onnx.load(onnx_file)
model_simp, check = simplify(model_onnx2)
onnx.save(model_simp, onnx_file)

```
![image](https://user-images.githubusercontent.com/33194443/218127198-5229ca32-025d-49bb-9aaa-2beeafaf6817.png)

**Comment by PINTO0309** [comment]
```python
import torch
import torch.nn as nn


class Model(nn.Module):
    def __init__(self):
        super(Model, self).__init__()

    def forward(self, x):
        res = torch.reshape(x, [1,-1,3072])
        reduce, _ = torch.max(res, dim=2, keepdim=True)
        sub = torch.subtract(res, reduce)
        sub = sub + 1.0
        ret = torch.softmax(sub, dim=2)
        return ret

model = Model()

x = torch.randn([1,23,64,48])
onnx_file = f'reducemax_softmax_workaround.onnx'
torch.onnx.export(
    model,
    args=(x),
    f=onnx_file,
    opset_version=11,
    input_names=[
        'input',
    ],
    output_names=[
        'output',
    ],
)
import onnx
from onnxsim import simplify
model_onnx2 = onnx.load(onnx_file)
model_simp, check = simplify(model_onnx2)
onnx.save(model_simp, onnx_file)
```
![image](https://user-images.githubusercontent.com/33194443/218127912-dd57036b-ae80-4a74-ae5f-5ab84719b03e.png)

**Comment by PINTO0309** [comment]
```python
import torch
import torch.nn as nn


class Model(nn.Module):
    def __init__(self):
        super(Model, self).__init__()

    def forward(self, x):
        res = torch.reshape(x, [1,-1,3072])
        reduce, _ = torch.max(res, dim=2, keepdim=True)
        sub = torch.subtract(res, reduce)
        sub = sub + 1e-16
        ret = torch.softmax(sub, dim=2)
        return ret

model = Model()

x = torch.randn([1,23,64,48])
onnx_file = f'reducemax_softmax_workaround.onnx'
torch.onnx.export(
    model,
    args=(x),
    f=onnx_file,
    opset_version=11,
    input_names=[
        'input',
    ],
    output_names=[
        'output',
    ],
)
import onnx
from onnxsim import simplify
model_onnx2 = onnx.load(onnx_file)
model_simp, check = simplify(model_onnx2)
onnx.save(model_simp, onnx_file)
```
![image](https://user-images.githubusercontent.com/33194443/218128339-0a7f1894-2e60-4683-a36f-40da56d07de0.png)

**Comment by PINTO0309** [comment]
What you could do now would be to insert meaningless addition and subtraction before `Softmax` to work around the TensorFlow bug.

```python
import torch
import torch.nn as nn

class Model(nn.Module):
    def __init__(self):
        super(Model, self).__init__()

    def forward(self, x):
        res = torch.reshape(x, [1,-1,3072])
        reduce, _ = torch.max(res, dim=2, keepdim=True)
        sub = torch.subtract(res, reduce)
        sub = sub + 1e-16
        sub = sub - 1e-16
        ret = torch.softmax(sub, dim=2)
        return ret

model = Model()

x = torch.randn([1,23,64,48])
onnx_file = f'reducemax_softmax_workaround.onnx'
torch.onnx.export(
    model,
    args=(x),
    f=onnx_file,
    opset_version=11,
    input_names=[
        'input',
    ],
    output_names=[
        'output',
    ],
)
import onnx
from onnxsim import simplify
model_onnx2 = onnx.load(onnx_file)
model_simp, check = simplify(model_onnx2)
onnx.save(model_simp, onnx_file)
```
![image](https://user-images.githubusercontent.com/33194443/218128710-2e00b0b7-5aa8-4db4-a06f-7abb99fbba62.png)

**Comment by PINTO0309** [comment]
This bug in TensorFlow is a very interesting behavior, which `torch.min()` does not reproduce.
```python
import torch
import torch.nn as nn


class Model(nn.Module):
    def __init__(self):
        super(Model, self).__init__()

    def forward(self, x):
        res = torch.reshape(x, [1,-1,3072])
        reduce, _ = torch.min(res, dim=2, keepdim=True)
        sub = torch.subtract(res, reduce)
        # sub = sub + 1e-16
        # sub = sub - 1e-16
        sub = sub * 1.0
        ret = torch.softmax(sub, dim=2)
        return ret

model = Model()

x = torch.randn([1,23,64,48])
onnx_file = f'reducemax_softmax_workaround.onnx'
torch.onnx.export(
    model,
    args=(x),
    f=onnx_file,
    opset_version=11,
    input_names=[
        'input',
    ],
    output_names=[
        'output',
    ],
)
import onnx
from onnxsim import simplify
model_onnx2 = onnx.load(onnx_file)
model_simp, check = simplify(model_onnx2)
onnx.save(model_simp, onnx_file)
```
![image](https://user-images.githubusercontent.com/33194443/218229154-c6110a44-e96f-474f-b4a9-4fc54c2de9af.png)

**Comment by PINTO0309** [comment]
```python
import torch
import torch.nn as nn


class Model(nn.Module):
    def __init__(self):
        super(Model, self).__init__()

    def forward(self, x):
        res = torch.reshape(x, [1,-1,3072])
        # reduce, _ = torch.max(res, dim=2, keepdim=True)
        reduce = torch.mean(res, dim=2, keepdim=True)
        sub = torch.subtract(res, reduce)
        # sub = sub + 1e-16
        # sub = sub - 1e-16
        # sub = sub * 1.0
        ret = torch.softmax(sub, dim=2)
        return ret

model = Model()

x = torch.randn([1,23,64,48])
onnx_file = f'reducemax_softmax_workaround.onnx'
torch.onnx.export(
    model,
    args=(x),
    f=onnx_file,
    opset_version=11,
    input_names=[
        'input',
    ],
    output_names=[
        'output',
    ],
)
import onnx
from onnxsim import simplify
model_onnx2 = onnx.load(onnx_file)
model_simp, check = simplify(model_onnx2)
onnx.save(model_simp, onnx_file)
```
![image](https://user-images.githubusercontent.com/33194443/218230159-f349ede4-11b2-4a95-9c92-3a2c399e00e1.png)

**Comment by PINTO0309** [comment]
It seems that TensorFlow only behaves incorrectly when processing `torch.max()` -> `torch.subtract()` -> `torch.softmax()` in that order.
```python
import torch
import torch.nn as nn


class Model(nn.Module):
    def __init__(self):
        super(Model, self).__init__()

    def forward(self, x):
        res = torch.reshape(x, [1,-1,3072])
        reduce, _ = torch.max(res, dim=2, keepdim=True)
        # reduce = torch.mean(res, dim=2, keepdim=True)
        # sub = torch.subtract(res, reduce)
        sub = torch.add(res, reduce)
        # sub = sub + 1e-16
        # sub = sub - 1e-16
        # sub = sub * 1.0
        ret = torch.softmax(sub, dim=2)
        return ret

model = Model()

x = torch.randn([1,23,64,48])
onnx_file = f'reducemax_softmax_workaround.onnx'
torch.onnx.export(
    model,
    args=(x),
    f=onnx_file,
    opset_version=11,
    input_names=[
        'input',
    ],
    output_names=[
        'output',
    ],
)
import onnx
from onnxsim import simplify
model_onnx2 = onnx.load(onnx_file)
model_simp, check = simplify(model_onnx2)
onnx.save(model_simp, onnx_file)
```
![image](https://user-images.githubusercontent.com/33194443/218230629-1e49f1ae-ddd7-4265-bfcb-4816aef81127.png)

**Comment by PINTO0309** [comment]
https://github.com/PINTO0309/onnx2tf/releases/tag/1.6.4

**Comment by H4dr1en** [comment]
Thanks for investigating further! The workaround solved my problem 👍
